### PR TITLE
[core] Eliminate __udivdi3 in millis() on ESP32 and RP2040

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

On 32-bit targets (ESP32 Xtensa/RISC-V, RP2040 Cortex-M0+), GCC does not optimize 64-bit constant division (`/ 1000ULL`) into a multiply-by-reciprocal. Instead it emits a call to `__udivdi3` — a generic software 64-bit division routine.

`millis()` is called once per looping component per iteration (via `WarnIfComponentBlockingGuard::finish()`), plus overhead calls. A typical device with ~21 components makes ~21 calls/loop; a complex device with ~120 components makes ~120 calls/loop.

This PR adds a templated `micros_to_millis<ReturnT>()` helper that eliminates the 64-bit division in both `millis()` and `millis_64()` by exploiting `1000 = 8 × 125`:

1. **Right-shift by 3** — free divide-by-8
2. **[Euclidean decomposition](https://en.wikipedia.org/wiki/Euclidean_division)** of the remaining ÷125, splitting the 64-bit value into 32-bit hi/lo words:

```
2^32 = 34359738 × 125 + 46
(hi × 2^32 + lo) / 125 = hi × 34359738 + (hi × 46 + lo) / 125
```

GCC optimizes the final 32-bit `/ 125U` into a multiply-by-reciprocal (`mulhu` + `srli`), so no division instruction is emitted.

The template parameter controls the return type:
- `micros_to_millis(us)` → `uint32_t` (default, for `millis()`) — truncated `hi * Q`
- `micros_to_millis<uint64_t>(us)` → `uint64_t` (for `millis_64()`) — widened 32×32→64 `hi * Q`

Using D=125 (R=46) instead of D=1000 (R=296) makes `hi × R` cheaper and extends the safe input range 51× wider (~3.2×10¹⁸ vs ~6.2×10¹⁶ microseconds, i.e. ~101,700 years vs ~1,975 years).

The function is `constexpr`, enabling compile-time validation via `static_assert`.

### Benchmark (ESP32 classic, Xtensa LX6, 240 MHz)

100k iterations per measurement:

**32-bit result (`millis()`):**

| Scenario | `__udivdi3` | `micros_to_millis` | Speedup |
|---|---|---|---|
| Fresh boot (hi=0) | 647 ns | 271 ns | **2.39×** |
| 30-day uptime (hi=603) | 696 ns | 277 ns | **2.52×** |
| Carry path (overflow) | 712 ns | 258 ns | **2.76×** |

**64-bit result (`millis_64()`):**

| Scenario | `__udivdi3` | `micros_to_millis<uint64_t>` | Speedup |
|---|---|---|---|
| Fresh boot (hi=0) | 708 ns | 377 ns | **1.88×** |
| 30-day uptime (hi=603) | 771 ns | 377 ns | **2.04×** |
| Carry path (overflow) | 778 ns | 377 ns | **2.06×** |

Estimated savings @ 7100 loops/min (32-bit `millis()` calls only):

| Device complexity | Calls/loop | Savings/loop | Savings/sec | Savings/min |
|---|---|---|---|---|
| Typical (~21 components) | ~21 | ~9 µs | ~1.1 ms | ~64 ms |
| Complex (~120 components) | ~120 | ~54 µs | ~6.4 ms | ~384 ms |

### Correctness

Verified against reference `us // 1000` across **140 unit test cases** covering both 32-bit and 64-bit variants: boundary values, hi/lo combinations, carry-path overflow, realistic uptimes (30 days, 1 year, safe limit), shift-boundary mod-8 variations, and upper-bit preservation for the 64-bit result.

Additionally verified with a standalone C++ test across **124M+ values** — zero mismatches. See attached [test_fast_div1000_32.zip](https://github.com/user-attachments/files/25678044/test_fast_div1000_32.zip).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

N/A — performance optimization

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

N/A — no user-facing changes

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

Benchmark button to reproduce the measurements:

```yaml
button:
  - platform: template
    name: "Benchmark millis()"
    on_press:
      - lambda: |-
          // shift3+div125: 1000 = 8 * 125, shift then Euclidean decompose /125
          // 2^32 = 34359738*125 + 46
          auto millis_fast = [](uint64_t us) -> uint32_t {
            uint64_t x = us >> 3;
            uint32_t lo = (uint32_t) x;
            uint32_t hi = (uint32_t) (x >> 32);
            uint32_t adj = hi * 46U + lo;
            return hi * 34359738U + (adj < lo ? (adj + 46U) / 125U + 34359738U : adj / 125U);
          };

          // 64-bit result variant (for millis_64)
          auto millis_fast_64 = [](uint64_t us) -> uint64_t {
            uint64_t x = us >> 3;
            uint32_t lo = (uint32_t) x;
            uint32_t hi = (uint32_t) (x >> 32);
            uint32_t adj = hi * 46U + lo;
            return (uint64_t)hi * 34359738U + (adj < lo ? (adj + 46U) / 125U + 34359738U : adj / 125U);
          };

          auto div64_current = [](uint64_t us) -> uint32_t {
            return (uint32_t)(us / 1000ULL);
          };

          auto div64_64bit = [](uint64_t us) -> uint64_t {
            return us / 1000ULL;
          };

          const uint32_t ITERATIONS = 100000;
          volatile uint32_t sink = 0;
          volatile uint64_t sink64 = 0;

          uint64_t base = millis_64() * 1000ULL;
          uint64_t base_aged = base + 2592000000000ULL;  // 30 days

          // 32-bit correctness + benchmark
          ESP_LOGI("bench", "div64=%u fast=%u match=%s",
                   div64_current(base_aged), millis_fast(base_aged),
                   div64_current(base_aged) == millis_fast(base_aged) ? "YES" : "NO");

          // 64-bit correctness + benchmark
          ESP_LOGI("bench", "div64_64=%" PRIu64 " fast_64=%" PRIu64 " match=%s",
                   div64_64bit(base_aged), millis_fast_64(base_aged),
                   div64_64bit(base_aged) == millis_fast_64(base_aged) ? "YES" : "NO");

          uint32_t t0 = micros();
          for (uint32_t i = 0; i < ITERATIONS; i++) sink = div64_current(base_aged + i);
          float ns_div32 = (float)(micros() - t0) / ITERATIONS * 1000.0f;

          t0 = micros();
          for (uint32_t i = 0; i < ITERATIONS; i++) sink = millis_fast(base_aged + i);
          float ns_fast32 = (float)(micros() - t0) / ITERATIONS * 1000.0f;

          t0 = micros();
          for (uint32_t i = 0; i < ITERATIONS; i++) sink64 = div64_64bit(base_aged + i);
          float ns_div64 = (float)(micros() - t0) / ITERATIONS * 1000.0f;

          t0 = micros();
          for (uint32_t i = 0; i < ITERATIONS; i++) sink64 = millis_fast_64(base_aged + i);
          float ns_fast64 = (float)(micros() - t0) / ITERATIONS * 1000.0f;

          ESP_LOGI("bench", "32-bit: div=%.0fns fast=%.0fns (%.1fx)", ns_div32, ns_fast32, ns_div32/ns_fast32);
          ESP_LOGI("bench", "64-bit: div=%.0fns fast=%.0fns (%.1fx)", ns_div64, ns_fast64, ns_div64/ns_fast64);
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by bdraco** [comment]
[test_fast_div1000_32.zip](https://github.com/user-attachments/files/25678044/test_fast_div1000_32.zip)

**Comment by bdraco** [comment]
[test_fast_div1000_32.zip](https://github.com/user-attachments/files/25678244/test_fast_div1000_32.zip)

**Comment by bdraco** [comment]
```
[23:12:31.389][D][button:019]: 'Benchmark millis()' Pressed.
[23:12:31.392][I][bench:098]: --- FRESH (hi=0): base_us=9999000 hi=0 lo=9999000 ---
[23:12:31.396][I][bench:105]:   check: div64=9999 onediv=9999 ternary=9999 match=YES
[23:12:31.682][I][bench:141]:   micros() [timer read]:      754.4 ns
[23:12:31.683][I][bench:142]:   millis() [current impl]:    817.7 ns
[23:12:31.683][I][bench:143]:   __udivdi3 only:             671.8 ns
[23:12:31.683][I][bench:144]:   onediv (if/else):           335.4 ns
[23:12:31.687][I][bench:145]:   onediv (ternary):           257.0 ns
[23:12:31.687][I][bench:146]:   speedup vs div64: if=2.00x ternary=2.61x
[23:12:31.689][I][bench:098]: --- AGED (hi>0): base_us=2592009999000 hi=603 lo=2144719512 ---
[23:12:31.697][I][bench:105]:   check: div64=2592009999 onediv=2592009999 ternary=2592009999 match=YES
[23:12:32.013][I][bench:141]:   micros() [timer read]:      759.7 ns
[23:12:32.013][I][bench:142]:   millis() [current impl]:    817.9 ns
[23:12:32.013][I][bench:143]:   __udivdi3 only:             710.7 ns
[23:12:32.013][I][bench:144]:   onediv (if/else):           301.8 ns
[23:12:32.013][I][bench:145]:   onediv (ternary):           245.1 ns
[23:12:32.013][I][bench:146]:   speedup vs div64: if=2.36x ternary=2.90x
[23:12:32.013][I][bench:098]: --- CARRY (overflow): base_us=2594160242688 hi=603 lo=4294963200 ---
[23:12:32.013][I][bench:105]:   check: div64=2594160242 onediv=2594160242 ternary=2594160242 match=YES
[23:12:32.287][I][bench:141]:   micros() [timer read]:      759.1 ns
[23:12:32.289][I][bench:142]:   millis() [current impl]:    821.4 ns
[23:12:32.289][I][bench:143]:   __udivdi3 only:             714.2 ns
[23:12:32.295][I][bench:144]:   onediv (if/else):           300.5 ns
[23:12:32.295][I][bench:145]:   onediv (ternary):           245.9 ns
[23:12:32.295][I][bench:146]:   speedup vs div64: if=2.38x ternary=2.90x
```

**Comment by bdraco** [comment]
[test_fast_div1000_32.zip](https://github.com/user-attachments/files/25678905/test_fast_div1000_32.zip)

File: esphome/core/helpers.h [hunk 1]
```diff
@@ -527,6 +527,36 @@ template<std::integral T> constexpr uint32_t fnv1a_hash_extend(uint32_t hash, T
...
+inline ESPHOME_ALWAYS_INLINE uint32_t fast_div1000_32(uint64_t us) {
+  static constexpr uint32_t D = 1000U;     // divisor (microseconds per millisecond)
+  static constexpr uint32_t Q = 4294967U;  // 2^32 / 1000
+  static constexpr uint32_t R = 296U;      // 2^32 % 1000
+  uint32_t lo = static_cast<uint32_t>(us);
+  uint32_t hi = static_cast<uint32_t>(us >> 32);
+  // Combine remainder term: hi * (2^32 % 1000) + lo
+  uint32_t adj = hi * R + lo;
+  // If adj overflowed, the true value is 2^32 + adj; apply the identity again
+  return hi * Q + (adj < lo ? (adj + R) / D + Q : adj / D);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/helpers.h]: `hi * R` cannot overflow for any realistic uptime:

- `esp_timer_get_time()` returns microseconds as `int64_t`
- At **2 years** uptime: `hi` ≈ 14,685, so `hi * R` = 14,685 × 296 = **4,346,760** (well within uint32_t)
- `hi` only reaches `UINT32_MAX / 296` ≈ 14.5 million at **~584,942 years** of uptime
- The RP2040 `time_us_64()` has the same range

The suggested fix reintroduces `adj64 / D` — a 64-bit division that calls `__udivdi3`, which is the exact thing this PR eliminates.

**Comment by bdraco** [comment]
[test_fast_div1000_32.zip](https://github.com/user-attachments/files/25680279/test_fast_div1000_32.zip)

**Comment by bdraco** [comment]
# Remaining `__udivdi3` call sites (cold paths)

These 64-bit divisions still emit `__udivdi3` on 32-bit targets but were intentionally left unfixed as none are hot paths.

| File | Line | Expression | Frequency |
|------|------|------------|-----------|
| `uptime/text_sensor/uptime_text_sensor.cpp` | 22 | `millis_64() / 1000` | Once per update interval (default 60s) |
| `uptime/sensor/uptime_seconds_sensor.cpp` | 12-13 | `uptime / 1000ULL` and `uptime % 1000ULL` | Once per update interval |
| `web_server/web_server.cpp` | 388, 417 | `millis_64() / 1000` | On HTTP request only |

Only millis/millis_64 is hot so the other are left as-is with a note here in case they become hot

File: esphome/core/helpers.h [hunk 2]
```diff
@@ -599,6 +599,48 @@ template<std::integral T> constexpr uint32_t fnv1a_hash_extend(uint32_t hash, T
...
+  constexpr uint32_t r = static_cast<uint32_t>((1ULL << 32) % d);  // 46
+  // 1000 = 8 * 125; divide-by-8 is a free shift
+  uint64_t x = us >> 3;
+  uint32_t lo = static_cast<uint32_t>(x);
+  uint32_t hi = static_cast<uint32_t>(x >> 32);
+  // Combine remainder term: hi * (2^32 % 125) + lo
+  uint32_t adj = hi * r + lo;
+  // If adj overflowed, the true value is 2^32 + adj; apply the identity again
+  // static_cast<ReturnT>(hi) widens to 64-bit when ReturnT=uint64_t, preserving upper bits of hi*q
+  return static_cast<ReturnT>(hi) * q + (adj < lo ? (adj + r) / d + q : adj / d);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/helpers.h:637]: its an internal helper.. I think 100000+ years of uptime is unlikely

File: tests/unit_tests/test_micros_to_millis.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+@pytest.mark.parametrize("us", BOUNDARY_VALUES, ids=lambda v: f"us={v}")
+def test_64bit_boundary_values(us: int) -> None:
+    assert micros_to_millis_64(us) == reference_64(us)
+
+
+@pytest.mark.parametrize("hi", HI_VALUES, ids=lambda v: f"hi={v}")
+@pytest.mark.parametrize("lo_offset", LO_VALUES, ids=lambda v: f"lo={v}")
+def test_64bit_hi_lo_combinations(hi: int, lo_offset: int) -> None:
+    us = (hi << 32) | lo_offset
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/unit_tests/test_micros_to_millis.py]: 15a792ecce

File: tests/unit_tests/test_micros_to_millis.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+@pytest.mark.parametrize("us", BOUNDARY_VALUES, ids=lambda v: f"us={v}")
+def test_32bit_boundary_values(us: int) -> None:
+    assert micros_to_millis(us) == reference_32(us)
+
+
+@pytest.mark.parametrize("hi", HI_VALUES, ids=lambda v: f"hi={v}")
+@pytest.mark.parametrize("lo_offset", LO_VALUES, ids=lambda v: f"lo={v}")
+def test_32bit_hi_lo_combinations(hi: int, lo_offset: int) -> None:
+    us = (hi << 32) | lo_offset
+    assert micros_to_millis(us) == reference_32(us)
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/unit_tests/test_micros_to_millis.py]: 15a792ecce

**Comment by bdraco** [comment]
```
[09:35:12.970][D][button:019]: 'Log millis' Pressed.
[09:35:12.974][I][millis_test:058]: millis()=100884  millis_64()=100884
[09:35:13.175][D][button:019]: 'Log millis' Pressed.
[09:35:13.179][I][millis_test:058]: millis()=101091  millis_64()=101091
[09:35:13.402][D][button:019]: 'Log millis' Pressed.
[09:35:13.404][I][millis_test:058]: millis()=101317  millis_64()=101317
[09:35:13.584][D][button:019]: 'Log millis' Pressed.
[09:35:13.588][I][millis_test:058]: millis()=101499  millis_64()=101499
[09:35:13.769][D][button:019]: 'Log millis' Pressed.
[09:35:13.772][I][millis_test:058]: millis()=101684  millis_64()=101684

```

 euclidean decomposition implementation is producing identical result

**Comment by bdraco** [comment]
pr failure is github being flakey.. close reopen

**Review by swoboda1337 — APPROVED** [review]
Yeah the math looks correct.

**Comment by bdraco** [comment]
Thanks
